File: twoSum.py
from typing import List


# 不知道是哪个
class Solution1:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        nums = sorted([(nums[i], i) for i in range(len(nums))])
        i = 0
        j = len(nums) - 1

        while i < j:
            a = nums[i][0] + nums[j][0]

            if a == target:
                return [nums[i][1], nums[j][1]]
            elif a > target:
                j -= 1
            else:
                i += 1

        return [-1, -1]


# nums = [2, 7, 11, 15]
# target = 9
# nums = [3, 2, 4]
# target = 6
# nums = [3, 2, 4]
# target = 6
# print(Solution().twoSum(nums, target))


# 167. 两数之和 II - 输入有序数组
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        numbers = sorted([(numbers[i], i) for i in range(len(numbers))])
        i = 0
        j = len(numbers) - 1

        while i < j:
            a = numbers[i][0] + numbers[j][0]

            if a == target:
                return [i + 1, j + 1]
            elif a > target:
                j -= 1
            else:
                i += 1

        return [-1, -1]


numbers = [2, 7, 11, 15]
target = 9
numbers = [2, 3, 4]
target = 6
print(Solution().twoSum(numbers, target))

File: twoSum.cpp
/**
 * @author f4prime
 * @email zzz136454872@163.com
 * @aim a plain model for leetcode
 */

#include <algorithm>
#include <iostream>
#include <map>
#include <queue>
#include <stack>
#include <string>
#include <vector>

using namespace std;

//显示中间结果
void print(vector<int> a) {
  for (int i = 0; i < (int)a.size(); i++)
    cout << a[i] << " ";
  cout << endl;
}

void print(vector<vector<int>> a) {
  for (int i = 0; i < (int)a.size(); i++) {
    for (int j = 0; j < (int)a[i].size(); j++)
      cout << a[i][j] << " ";
    cout << endl;
  }
}

void print(vector<bool> a) {
  for (int i = 0; i < (int)a.size(); i++) {
    cout << i << " ";
    if (a[i])
      cout << "true ";
    else
      cout << "false ";
  }
  cout << endl;
}

void print(bool a[], int len) {
  for (int i = 0; i < len; i++) {
    cout << i << " ";
    if (a[i])
      cout << "true ";
    else
      cout << "false ";
  }
  cout << endl;
}

//#define testMod

#ifdef testMod
void test() {}
#endif

#ifndef testMod
class Solution {
public:
  vector<int> twoSum(vector<int> &numbers, int target) {
    int i = 0, j = numbers.size() - 1;
    vector<int> out;
    int tmp;
    while (i < j) {
      tmp = numbers[i] + numbers[j];
      if (tmp == target) {
        out.push_back(i + 1);
        out.push_back(j + 1);
        return out;
      } else if (tmp > target)
        j--;
      else
        i++;
    }
    return out;
  }

private:
};
#endif

int main() {
#ifdef testMod
  test();
#endif

#ifndef testMod
  Solution sl;

#endif
  return 0;
}
